test/jdk/java/net/httpclient/ContentLengthHeaderTest.java

```diff
@@ -0,0 +1,229 @@
+ * questions.
+ */
+
+/*
+ * @test
+ * @summary Tests that a Content-length header is not sent for GET requests
+ *          that do not have a body. Also checks that the header is sent when
+ *          a body is present on the GET request.
+ * @library /test/lib
+ * @bug 8283544
+ * @run testng/othervm ContentLengthHeaderTest
+ */
+
+
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
+import org.testng.annotations.AfterTest;
+import org.testng.annotations.BeforeTest;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import jdk.test.lib.net.URIBuilder;
+
+
+import static java.net.http.HttpClient.Version.HTTP_1_1;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.testng.Assert.assertEquals;
+
+
+public class ContentLengthHeaderTest {
+
+    final String NO_BODY_PATH = "/no_body";
+    final String BODY_PATH = "/body";
+
+    static HttpServer testContentLengthServer;
+    static PrintStream testLog = System.err;
+
+    URI testContentLengthURI;
+
+    @BeforeTest
+    public void setup() throws IOException, URISyntaxException {
+        InetSocketAddress sa = new InetSocketAddress(InetAddress.getLoopbackAddress(), 0);
+        testContentLengthServer = HttpServer.create(sa, 0);
+
+        // Create handlers for tests that check for the presence of a Content-length header
+        testContentLengthServer.createContext(NO_BODY_PATH, new NoContentLengthHandler());
+        testContentLengthServer.createContext(BODY_PATH, new ContentLengthHandler());
+        testContentLengthURI = URIBuilder.newBuilder()
+                .scheme("http")
+                .loopback()
+                .port(testContentLengthServer.getAddress().getPort())
+                .build();
+
+        testContentLengthServer.start();
+        testLog.println("Server up at address: " + testContentLengthServer.getAddress());
+        testLog.println("Request URI for Client: " + testContentLengthURI);
+    }
+
+    @AfterTest
+    public void teardown() {
+        if (testContentLengthServer != null)
+            testContentLengthServer.stop(0);
+    }
+
+    @Test
+    // A GET request with no request body should have no Content-length header
+    public void getWithNoBody() throws IOException, InterruptedException {
+        testLog.println("Checking GET with no request body");
+        HttpRequest req = HttpRequest.newBuilder()
+                .version(HTTP_1_1)
+                .GET()
+                .uri(URI.create(testContentLengthURI + NO_BODY_PATH))
+                .build();
+        HttpClient hc = HttpClient.newHttpClient();
```

Comment:
It would be better to reuse the same client for all tests. Also better to configure it with NO_PROXY.